community/rfcs/techniques/proposals/20200205-weight-clustering-proposal.md

```diff
@@ -0,0 +1,83 @@
+# Keras Weight Clustering API
+
+| Status        | Pending                                       |
+:---------------|:----------------------------------------------|
+| **Author(s)** | Mohamed Nour Abouelseoud (mohamednour.abouelseoud@arm.com), Aron Virginas-Tar (aron.virginas-tar@arm.com), Anton Kachatkou (anton.kachatkou@arm.com)|
+| **Sponsor**   | Alan Chiao (alanchiao@google.com)             |
+| **Updated**   | 2020-03-16                                    |
+
+We propose to implement the technique of weight clustering on top of the Keras API.
+
+## Motivation and Overview
+
+Clustering, or weight sharing, brings improvements via model compression. The technique revolves around reducing the number of unique weights in a model, thus facilitating storage and over-the-wire size reduction. Further framework and hardware support can also unlock memory footprint improvements.
+
+Clustering can be combined with pruning and quantization, reducing the network’s storage requirements even further &mdash; see the paper linked [below](#proposed-design) for more details.
+
+## Experimental Results
+
+The table below summarizes the compression rate and top-1 accuracy achieved when clustering the weights of different models using linear centroid initialization.
+
+NOTE: Further networks, e.g. MobileNetV1, are being investigated. The table will be extended to include these once the results become available.
+
+<table>
+  <tr>
+    <th rowspan=2>Model</th>
+    <th rowspan=2>Dataset</th>
+    <th colspan=2>Original</th>
+    <th colspan=3>Clustered</th>
+  </tr>
+  <tr>
+    <th>Top-1 accuracy</th>
+    <th>Size of compressed .tflite</th>
+    <th># of clusters</th>
+    <th>Top-1 accuracy</th>
+    <th>Size of compressed .tflite</th>
+  </tr>
+  <tr>
+    <td>ConvNet</td>
+    <td>MNIST</td>
+    <td>99.40%</td>
+    <td>0.6 MB</td>
+    <td>32</td>
+    <td>98.78%</td>
+    <td>0.1 MB</td>
+  </tr>
+  <tr>
+    <td>MobileNetV2</td>
+    <td>ImageNet</td>
+    <td>72.29%</td>
+    <td>13.0 MB</td>
+    <td>32</td>
+    <td>69.33%</td>
+    <td>2.6 MB</td>
+  </tr>
+  <tr>
+    <td>DS-CNN-L</td>
+    <td>Speech Commands v0.02</td>
+    <td>95.03%</td>
+    <td>1.5 MB</td>
+    <td>32</td>
+    <td>94.71%</td>
+    <td>0.3 MB</td>
+  </tr>
+</table>
+
+NOTE: *Size of compressed .tflite* refers to the size of the zipped .tflite file obtained from the model through the following process:
+1. Serialize the Keras model into .h5 file
```

Comment:
Note this should also apply to the zipped SavedModel, correct?

Reply:
Yes, but the improvement in weight compression may be masked by other data stored within SavedModel.